[group key addrs 3/?]: send and receive support for V2 addresses

guggero opened the pull request:
Depends on https://github.com/lightninglabs/taproot-assets/pull/1502 and https://github.com/lightninglabs/taproot-assets/pull/1512.

This is a very early prototype that barely works.
But it shows all the areas that need to be touched to support V2 addresses.

This will likely be split into two or three PRs, once fully finished and cleaned up.

guggero commented on proof/send.go:
```diff
@@ -0,0 +1,244 @@
+	// ScriptKeyDerivationUniquePedersen means the script key is derived
+	// using the address's recipient ID key and a single leaf that contains
+	// an un-spendable Pedersen commitment key
+	// (OP_CHECKSIG <NUMS_key + asset_id * G>). This can be used to
```
Correct. Will add the odd TLV value to make sure identical sends (e.g. same amounts and same asset IDs) will still produce a different on-chain output for better privacy.

guggero commented on proof/send.go:
```diff
@@ -0,0 +1,244 @@
+	// TxProof is the proof of the transaction that contains the asset
+	// outputs that are being sent. This is used as proof-of work to show
+	// to the auth mailbox server.
+	TxProof TxProof
```
The envelope itself isn't encoded, only the `SendFragment` within. So the `TxProof` here is is what's required to send the fragment to the server.
Perhaps "envelope" isn't the best analogy here, as that's also sent along with a letter/message. Going to rename this to `SendManifest` instead, perhaps that makes more sense.

guggero commented on tapfreighter/chain_porter.go:
```diff
@@ -882,6 +922,29 @@ func (p *ChainPorter) transferReceiverProof(pkg *sendPackage) error {
+				"service handle: %w", err)
+		}
+
+		err = courier.DeliverFragment(ctx, envelope)
```
Good point. The cardinality is different for fragments vs. proofs. But I made it so that we can always try again to send the same message, so we just try in the same loop. Also makes the whole backoff and re-try logic work seamlessly.

guggero commented on tapcfg/server.go:
```diff
@@ -216,7 +216,7 @@ func genServerConfig(cfg *Config, cfgLogger btclog.Logger,
 		if cfg.DefaultProofCourierAddr == defaultProofCourierAddr {
 			cfg.DefaultProofCourierAddr = fmt.Sprintf(
-				"%s://%s", proof.UniverseRpcCourierType,
+				"%s://%s", proof.AuthMailboxUniRpcCourierType,
```
Yes, currently that is the case. I assume that's how we're going to roll out things on our end, at least for our own universe servers.

guggero commented on tapfreighter/wallet.go:
```diff
@@ -274,12 +274,187 @@ func (f *AssetWallet) FundAddressSend(ctx context.Context,
+			randAltLeaf, err := asset.NewAltLeaf(
+				randScriptKey, asset.ScriptV0,
+			)
+			vOut.AltLeaves = append(vOut.AltLeaves, randAltLeaf)
```
Yeah, I should've commented on this somewhere... So we create the split commitments from the virtual outputs. So any field we want to set on an asset also needs to be in the `VOutput` and therefore serialized in the `VPacket`. So it seemed a bit overkill to add the unknown TLV types map to the virtual output just for this. And we'd likely carry that data around forever (since we create copies of assets when we spend them).
So using an alt leaf felt nice and easy because that way we don't encumber the asset itself.
And because address related outputs are _always_ split outputs and we never check split outputs for anything STXO related this shouldn't really have an impact on that part of the code.

But happy to change things if you feel like this approach is the easy way out.

ZZiigguurraatt commented on taprpc/taprootassets.proto:
```diff
@@ -1374,12 +1409,17 @@ message AddrReceivesResponse {
     repeated string tap_addrs = 1;
 
     // The optional fee rate to use for the minting transaction, in sat/kw.
     uint32 fee_rate = 2;
```
Why do we not use the field name `sat_per_vbyte` like in `SendCoinsRequest` (https://github.com/lightningnetwork/lnd/blob/6b326152d419d0777f739a23c0d9330164d26c4d/lnrpc/lightning.proto#L1213-L1217)?

guggero replied:
Legacy reasons, should fix in separate PR.

ZZiigguurraatt replied:
Created https://github.com/lightninglabs/taproot-assets/issues/1661 to track this.

ZZiigguurraatt commented on taprpc/taprootassets.proto:
```diff
@@ -1374,12 +1409,17 @@ message AddrReceivesResponse {
 }
 
 message SendAssetRequest {
```
Do we want to add a `min_confs` and/or `spend_unconfirmed` option like in `SendCoinsRequest` (https://github.com/lightningnetwork/lnd/blob/6b326152d419d0777f739a23c0d9330164d26c4d/lnrpc/lightning.proto#L1231-L1238) or does no one use those options?

guggero replied:
`spend_unconfirmed` will never work with Taproot Assets, because to create a valid proof the transaction always has to be in a block. Or I guess not "never", just not without a major protocol enhancement/upgrade.

ZZiigguurraatt replied:
Created https://github.com/lightninglabs/taproot-assets/issues/1663 to track this

ZZiigguurraatt commented:
I need some guidance on what litd commit I can put this PR into. If I try to build from https://github.com/lightninglabs/lightning-terminal/commit/a0915a30462f1a06b9355154382fb883bcd9ba70 using the `TAPROOT_ASSETS_VERSION` build arg in `dev.Dockerfile` (https://github.com/lightninglabs/lightning-terminal/blob/a0915a30462f1a06b9355154382fb883bcd9ba70/dev.Dockerfile#L38C5-L38C27) I get

```
239.1 /go/pkg/mod/github.com/lightninglabs/taproot-assets@v0.6.1-0.20250710113743-2de283dbbb63/rpcutils/marshal.go:163:14: undefined: taprpc.ScriptKeyType_SCRIPT_KEY_UNIQUE_PEDERSEN
239.1 /go/pkg/mod/github.com/lightninglabs/taproot-assets@v0.6.1-0.20250710113743-2de283dbbb63/rpcutils/marshal.go:190:17: undefined: taprpc.ScriptKeyType_SCRIPT_KEY_UNIQUE_PEDERSEN
```

ZZiigguurraatt commented:
> I need some guidance on what litd commit I can put this PR into. If I try to build from [lightninglabs/lightning-terminal@a0915a3](https://github.com/lightninglabs/lightning-terminal/commit/a0915a30462f1a06b9355154382fb883bcd9ba70) using the `TAPROOT_ASSETS_VERSION` build arg in `dev.Dockerfile` (https://github.com/lightninglabs/lightning-terminal/blob/a0915a30462f1a06b9355154382fb883bcd9ba70/dev.Dockerfile#L38C5-L38C27) I get
> 
> ```
> 239.1 /go/pkg/mod/github.com/lightninglabs/taproot-assets@v0.6.1-0.20250710113743-2de283dbbb63/rpcutils/marshal.go:163:14: undefined: taprpc.ScriptKeyType_SCRIPT_KEY_UNIQUE_PEDERSEN
> 239.1 /go/pkg/mod/github.com/lightninglabs/taproot-assets@v0.6.1-0.20250710113743-2de283dbbb63/rpcutils/marshal.go:190:17: undefined: taprpc.ScriptKeyType_SCRIPT_KEY_UNIQUE_PEDERSEN
> ```

I'm also using the `NO_UI` build arg with litd, not sure if that is relevant?

guggero commented on address/address.go:
```diff
@@ -259,13 +292,39 @@ func CommitmentVersion(vers Version) (*commitment.TapCommitmentVersion,
+	// key as specified in the address directly.
+	if a.Version < V2 {
+		return &a.ScriptKey, nil
+	}
```
I was kind of assuming that any future address type would inherit all the properties of V2 addresses. So we would need to find and change fewer conditions. But I guess since it's hard to predict the future, we'd rather make it more explicit now.

guggero commented:
> Do we have an itest for testing non-grouped assets with the new V2 address format with and without an amount specified?

I've added an integration test for that in the latest push.

> This PR seems very big and I'd like to re-iterate the importance of https://github.com/lightninglabs/taproot-assets/issues/1492 .

I've added release notes (and the template) to the last commit. We'll need to back-fill this with PRs already merged for `v0.7.0`, will do that in a separate PR.

ZZiigguurraatt commented:
Please update https://github.com/lightninglabs/taproot-assets/blob/3bbfa95ea24207bba964945607aa2d4c369306d4/go.mod#L33 so that we are no longer on RC.

ZZiigguurraatt commented:
Can you update https://github.com/lightninglabs/taproot-assets/blob/3bbfa95ea24207bba964945607aa2d4c369306d4/go.mod#L131 to `github.com/lightningnetwork/lightning-onion v1.2.1-0.20240815225420-8b40adf04ab9` to match https://github.com/lightninglabs/taproot-assets/pull/1613/commits/8849fdbdcc21ad6b68bd57770b1804ba78ce33f1 ?

guggero commented:
> > > Please update
> > > https://github.com/lightninglabs/taproot-assets/blob/3bbfa95ea24207bba964945607aa2d4c369306d4/go.mod#L33
> > > 
> > > so that we are no longer on RC.
> > 
> > 
> > @ZZiigguurraatt merged here: #1657
> 
> you're saying this just needs a rebase then?

yes, on it

guggero commented on tapfreighter/chain_porter.go:
```diff
@@ -1458,6 +1458,25 @@ func (p *ChainPorter) stateStep(currentPkg sendPackage) (*sendPackage, error) {
 		// local means the lnd node connected to this daemon knows how
 		// to derive the key.
 		isLocalKey := func(key asset.ScriptKey) bool {
+			// To make sure we have the correct internal key with
+			// the family and index set, we attempt to fetch it
+			// from the database. If it exists, then we know we
+			// stored it with the correct information.
+			dbKey, err := p.cfg.AssetWallet.FetchScriptKey(
+				ctx, key.PubKey,
+			)
+			if err == nil {
+				return p.cfg.KeyRing.IsLocalKey(
+					ctx, dbKey.RawKey,
+				)
+			}
```
Okay, I added an error return value to the callback.

Yes, we need the fallback, in case the user only registered the internal key but not the outer script key.

guggero commented on tapdb/sqlc/queries/addrs.sql:
```diff
@@ -227,3 +227,12 @@ WHERE addr_events.status >= @status_from
+    ON addr_events.chain_txn_id = chain_txns.txn_id
+JOIN addrs
+    ON addr_events.addr_id = addrs.id
+WHERE addrs.version = $1;
```
Yes, that would've worked too. But the `max()` IMO makes the query a bit more explicit (even if the coalesce and cast make it a bit less readable).

guggero commented on tappsbt/decode.go:
```diff
@@ -524,3 +530,28 @@ func urlDecoder(u **url.URL) decoderFunc {
 	}
 }
+
+// addressDecoder returns a decoder function that can handle nil Taproot
+// addresses.
+func addressDecoder(a **address.Tap,
```
All the `xxxDecoder` and `xxxEncoder` functions in this file work the same way. I don't necessarily disagree with you but want to keep things consistent and not expand the diff by renaming all of the other ones too.

ZZiigguurraatt commented on docs/release-notes/release-notes-0.7.0.md:
```diff
@@ -0,0 +1,90 @@
+  are ideal as re-usable, long-term static addresses (with on-chain privacy
+  guarantees similar to BIP-0352 Silent Payments).
+  V2 addresses require the use of a proof courier that supports the
+  new `authmailbox+universerpc://` protocol. Any `tapd` that runs with version
```
I thought we decided we were going to make it auto discover the courier type, or do you want to create an issue to do that in a separate commit?

guggero replied:
It is auto-discovering, at least at the user-facing side. The user doesn't have to change or do anything. But the new protocol is still used internally and a universe server needs to support the protocol. So mentioning that in the release notes makes sense IMO, since the user will get an error message mentioning the protocol when they try to use a V2 address with a universe server that doesn't support it.

guggero commented:
Since https://github.com/lightninglabs/taproot-assets/issues/1665 relates to this not-yet-merged PR, I'm copying the comment here to not forget:

I get the following

```
fe99cea81980:/$ tapcli addrs new --group_key 023626a70077f8e7c9fe9598ad2d32b4d10ec326977750e4f88641c4c65d320815
[tapcli] unable to make addr: rpc error: code = Unknown desc = unable to make new addr: unable to make new addr: address: zero amount cannot be used for normal asset addresses of V0 or V1
fe99cea81980:/$ 
```

First off rather than tell us what address versions we cannot use, it would be better to tell us what versions we can use.

Even better, the CLI should be smart enough to know that only one address version is possible when using `group_key` and automatically use that version instead. When I use the gRPC call directly, I don't have to specify the `address_version`, it seems to automatically figure it out (as it should) and the CLI should too.

guggero commented:
Error message should be replace with one that is more informative for the user (https://github.com/lightninglabs/taproot-assets/issues/1666):

```
[tapcli] unable to make addr: rpc error: code = Unknown desc = unable to test connection to proof courier 'authmailbox+universerpc://alice:8443': unable to test connection proof courier 'authmailbox+universerpc://alice:8443': unable to connect to courier service during courier handle instantiation: unable to connect to auth mailbox RPC courier service: rpc error: code = Unimplemented desc = unknown service authmailboxrpc.Mailbox
```

guggero commented on tapfreighter/chain_porter.go:
```diff
@@ -28,6 +28,12 @@ import (
+const (
+	// DefaultSendFragmentExpiryDelta is the default number of blocks that
+	// we expect a send fragment to be valid for.
+	DefaultSendFragmentExpiryDelta = 210_000
```
Hmm, this comment made me think a lot more about backup and recovery procedures. I updated the main issue body of https://github.com/lightninglabs/taproot-assets/issues/426 quite a bit.

The conclusion I drew with respect to how long we should keep these messages is:
 - As long as the claimed outpoint of a message/proof (which, if used for our address v2 scheme represents the on-chain output the assets reside in) is unspent, we should _NOT_ delete any messages to assist in recovery.
 - Once the claimed outpoint of a message/proof was spent, after a grace period, we can assume the message is no longer relevant because its asset outputs have been spent. The duration of this grace period can be much shorter than the roughly 4 years proposed here. Perhaps roughly 90 days or 12'960 blocks after the output was spent?

Either way, I'm going to update the comments around the expiry to mention that this only applies to _spent_ outputs and is applied as a delta on top of the _spending_ transaction of the claimed outpoint.
No cleanup like that is implemented currently anyway, so the text should help us when we actually implement the cleanup.

guggero commented:
Okay, I've addressed all TODOs. And with the litd itest fixed, this should now also be testable on the `tapd-main-branch` of `litd`, cc @ZZiigguurraatt.

ZZiigguurraatt commented:
> Okay, I've addressed all TODOs. And with the litd itest fixed, this should now also be testable on the `tapd-main-branch` of `litd`, cc @ZZiigguurraatt.

still getting https://github.com/lightninglabs/taproot-assets/issues/1665

guggero commented on taprpc/authmailboxrpc/mailbox.proto:
```diff
@@ -107,9 +107,11 @@ message MailboxMessage {
-    // Block height after which this message may be considered expired by the
-    // system.
-    uint32 expiry_block_height = 4;
+    // A user-defined expiry block height delta for the message. Once the
+    // outpoint claimed by the proof provided by this message has been spent,
+    // then after this number of blocks after the spending transaction height
+    // the message can be considered expired and may be deleted.
+    uint32 expiry_block_delta = 4;
```
Hmm, you're right, this field currently makes no sense... I've removed it completely and will create an issue to look into cleaning up old auth mailbox messages instead.

guggero commented on tapfreighter/chain_porter.go:
```diff
@@ -29,6 +29,13 @@ import (
 	"github.com/lightningnetwork/lnd/lnwallet/chainfee"
 )
 
+const (
+	// DefaultSendFragmentExpiryDelta is the default number of blocks that
+	// we expect a send fragment to be valid for after its claimed outpoint
+	// has been spent. This is roughly equivalent to 90 days.
+	DefaultSendFragmentExpiryDelta = 12_960
+)
```
Removed.

guggero commented:
> still getting #1665 and #1666

@ZZiigguurraatt fixed both with the latest push. Make sure to also pull `tapd-main-branch`, as I had to force push that as well to fix this.

ZZiigguurraatt commented:
> > still getting #1665 and #1666
> 
> @ZZiigguurraatt fixed both with the latest push. Make sure to also pull `tapd-main-branch`, as I had to force push that as well to fix this.

This is not compiling for me from `tapd-main-branch`

```
157.6 # github.com/lightninglabs/loop/test
157.6 /go/pkg/mod/github.com/lightninglabs/loop@v0.31.2-beta/test/chainnotifier_mock.go:25:39: cannot use (*mockChainNotifier)(nil) (value of type *mockChainNotifier) as lndclient.ChainNotifierClient value in variable declaration: *mockChainNotifier does not implement lndclient.ChainNotifierClient (wrong type for method RegisterSpendNtfn)
157.6 		have RegisterSpendNtfn("golang.org/x/net/context".Context, *"github.com/btcsuite/btcd/wire".OutPoint, []byte, int32) (chan *chainntnfs.SpendDetail, chan error, error)
157.6 		want RegisterSpendNtfn("context".Context, *"github.com/btcsuite/btcd/wire".OutPoint, []byte, int32, ...lndclient.NotifierOption) (chan *chainntnfs.SpendDetail, chan error, error)
157.6 /go/pkg/mod/github.com/lightninglabs/loop@v0.31.2-beta/test/lnd_services_mock.go:44:19: cannot use chainNotifier (variable of type *mockChainNotifier) as lndclient.ChainNotifierClient value in struct literal: *mockChainNotifier does not implement lndclient.ChainNotifierClient (wrong type for method RegisterSpendNtfn)
157.6 		have RegisterSpendNtfn("golang.org/x/net/context".Context, *"github.com/btcsuite/btcd/wire".OutPoint, []byte, int32) (chan *chainntnfs.SpendDetail, chan error, error)
157.6 		want RegisterSpendNtfn("context".Context, *"github.com/btcsuite/btcd/wire".OutPoint, []byte, int32, ...lndclient.NotifierOption) (chan *chainntnfs.SpendDetail, chan error, error)
157.6 github.com/lightninglabs/loop/sweep
```



```
178.7 # github.com/lightninglabs/taproot-assets/authmailbox
178.7 /go/pkg/mod/github.com/lightninglabs/taproot-assets@v0.6.1-0.20250725201706-0e736bb0d7b1/authmailbox/client.go:167:3: unknown field ExpiryBlockHeight in struct literal of type authmailboxrpc.SendMessageRequest
178.7 /go/pkg/mod/github.com/lightninglabs/taproot-assets@v0.6.1-0.20250725201706-0e736bb0d7b1/authmailbox/marshal.go:31:3: unknown field ExpiryBlockHeight in struct literal of type authmailboxrpc.MailboxMessage
178.7 /go/pkg/mod/github.com/lightninglabs/taproot-assets@v0.6.1-0.20250725201706-0e736bb0d7b1/authmailbox/server.go:197:9: req.ExpiryBlockHeight undefined (type *authmailboxrpc.SendMessageRequest has no field or method ExpiryBlockHeight)
178.7 /go/pkg/mod/github.com/lightninglabs/taproot-assets@v0.6.1-0.20250725201706-0e736bb0d7b1/authmailbox/server.go:277:9: req.ExpiryBlockHeight undefined (type *authmailboxrpc.SendMessageRequest has no field or method ExpiryBlockHeight)
178.7 /go/pkg/mod/github.com/lightninglabs/taproot-assets@v0.6.1-0.20250725201706-0e736bb0d7b1/authmailbox/server.go:279:33: req.ExpiryBlockHeight undefined (type *authmailboxrpc.SendMessageRequest has no field or method ExpiryBlockHeight)
179.0 github.com/lightninglabs/taproot-assets/tappsbt
```

ZZiigguurraatt commented:
Please note: https://github.com/lightninglabs/taproot-assets/pull/1587#issuecomment-3132422101 is not related to https://github.com/lightninglabs/taproot-assets/pull/1587#issuecomment-3132335754 , I've tried restarting the network first.


```
ec56c799568c:/$ 
ec56c799568c:/$ tapcli assets balance --by_group
{
    "asset_balances":  {},
    "asset_group_balances":  {
        "022d7a190693f90c60e837a212b1e983a10d7ae9b1c8c16599ff5711695cd74c90":  {
            "group_key":  "022d7a190693f90c60e837a212b1e983a10d7ae9b1c8c16599ff5711695cd74c90",
            "balance":  "1649997"
        },
        "02e46c957cd201e759c1e2295313c9123965a0f1e7f8640b8bc4d797632bfd7fc6":  {
            "group_key":  "02e46c957cd201e759c1e2295313c9123965a0f1e7f8640b8bc4d797632bfd7fc6",
            "balance":  "2800000"
        },
        "0301f5f2a95b1ccd06d2168f3da5f457d75c7d6c1e056e66ee66c0b3a366742fb5":  {
            "group_key":  "0301f5f2a95b1ccd06d2168f3da5f457d75c7d6c1e056e66ee66c0b3a366742fb5",
            "balance":  "360000"
        }
    }
}
ec56c799568c:/$ 
ec56c799568c:/$ 
ec56c799568c:/$ 
ec56c799568c:/$ 
ec56c799568c:/$ tapcli assets send --addr taprt1qqqsyqspqqzzqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq5ssytt6ryrf87gvvr5r0gsjk85c8ggd0t5mrjxpvkvl74c3d9wdwnysqcssxf7x9snk7f9drz9egw2nhe6waf3swasvnn70hqamk89quxvctwsppqss99vsqasu8u6qvytfw3e9u7plr9fa4fv36rx06fc00hhf6xcx7flwpgqszrpyv96hg6rdv95kccn00q4h2mnfwejhyum9wfcxxw309askc6trv5arsdp5xvrza3rl
[tapcli] unable to send assets: rpc error: code = Unknown desc = error parsing addresses: addr 'taprt1qqqsyqspqqzzqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq5ssytt6ryrf87gvvr5r0gsjk85c8ggd0t5mrjxpvkvl74c3d9wdwnysqcssxf7x9snk7f9drz9egw2nhe6waf3swasvnn70hqamk89quxvctwsppqss99vsqasu8u6qvytfw3e9u7plr9fa4fv36rx06fc00hhf6xcx7flwpgqszrpyv96hg6rdv95kccn00q4h2mnfwejhyum9wfcxxw309askc6trv5arsdp5xvrza3rl' has a non-zero amount defined, cannot override amount in addresses_with_amounts
ec56c799568c:/$ 
```

guggero commented on address/address.go:
```diff
@@ -208,6 +233,16 @@ func New(version Version, genesis asset.Genesis, groupKey *btcec.PublicKey,
+	// Addresses with version 2 or later must use the new authmailbox proof
+	// courier type.
+	if version == V2 &&
+		proofCourierAddr.Scheme != proof.AuthMailboxUniRpcCourierType {
```
Yes, but this is the constructor of the address, so we don't have an instance yet to call the method on. So I decided to leave this in place but use methods for any logic used outside of the constructor.

guggero commented on tapgarden/custodian.go:
```diff
@@ -790,12 +874,43 @@ func (c *Custodian) mapToTapAddr(walletTx *lndclient.Transaction,
 	}
 
+	// V2 addresses aren't added to the internal wallet as an on-chain
+	// address, they can't be detected by watching the chain at all. We'll
+	// not mark them as managed in the address book, that way we'll always
+	// receive the full list of V2 addresses at startup as well.
```
The `AddressManaged` flag (though poorly named I admin) indicates that we have "processed" an address, meaning we have imported it into `lnd`'s wallet and will receive notifications for incoming on-chain transactions.
And on next startup, any address with the "managed" flag set to `true` will not be returned anymore.
But because we _always_ have to listen to any v2 addresses on the auth mailbox server, we don't want any v2 address to be set as managed (and we also don't import them into the `lnd` wallet as that doesn't make any sense).
I'll try to make the comment a bit more clear here.

guggero commented on tapgarden/custodian.go:
```diff
@@ -825,6 +940,221 @@ func (c *Custodian) importAddrToWallet(addr *address.AddrWithKeyInfo) error {
+		// Now that we've seen this output confirm on chain, we'll
+		// launch a goroutine to use the ProofCourier to import the
+		// proof into our local DB.
+		c.Wg.Add(1)
+		go func() {
+			defer c.Wg.Done()
```
Ah yes, that fits very nicely here. Thanks!

guggero commented:
I've addressed all of @ffranr's comments and also addressed https://github.com/lightninglabs/taproot-assets/pull/1587#issuecomment-3133714883, cc @ZZiigguurraatt.

guggero commented:
> I think we should attempt to make encoded V2 addresses smaller by dropping the zero asset ID requirement if we can. I don't think the BIP should hold us back, see review comment thread on this.

Will do that in a follow-up, tracking here: https://github.com/lightninglabs/taproot-assets/issues/1676

There seems to be an itest flake in Postgres, will look into that.

ZZiigguurraatt commented:
> No, for V2 addresses one should **only use the group key** for grouped assets. So I added an error message if the user tries to use `--asset_id` in combination with a V2 address for an asset that belongs to a group.

Looks good!

```
1953e99e3792:/$ tapcli addrs new --asset_id 072730e414d318f299faedb987b40ab5861c4e466003015d9e650e104355d597 --address_version v2
[tapcli] unable to make addr: rpc error: code = Unknown desc = unable to make new addr: version 2 addresses for grouped assets must use group key only
1953e99e3792:/$
```

guggero commented:
Found the itest issue, should be good to merge now.

Roasbeef reviewed (approved):
LGTM 🦜

> Reviewed 43 of 63 files at r10, 20 of 20 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/lightninglabs/taproot-assets/1587#-:-OWxAPlXEpo7svAFY-bj:b-lila8r)* status: all files reviewed, 60 unresolved discussions (waiting on @ffranr, @guggero, and @ZZiigguurraatt)

___
*[`itest/addrs_v2_test.go` line 50 at r11](https://reviewable.io/reviews/lightninglabs/taproot-assets/1587#-OWx72XU1ocgkECarV39:-OWx72XU1ocgkECarV3A:b-w5y6ih) ([raw file](https://github.com/lightninglabs/taproot-assets/blob/a45379ab9ff0951cf1f5ed926dcaa5b00ad97984/itest/addrs_v2_test.go#L50)):*
> ```Go
> 		amountToRequest := uint64(0)
> 		amountToSend := a.Amount - 1
> 		if amountToSend == 0 {
> ```

What type of error would get user get if they attempted to send a value greater than 1 to a collectible address? I wonder if we should encode that extra bit into the addr type. 

FWIW, calls to decode the address _should_ fetch the genesis as needed allowing this full information to be displayed. 
___
*[`tapdb/sqlc/queries/authmailbox.sql` line 52 at r6](https://reviewable.io/reviews/lightninglabs/taproot-assets/1587#-OWx7po1CsIdJkOEFT1Q:-OWx7po1CsIdJkOEFT1R:b-10it1l) ([raw file](https://github.com/lightninglabs/taproot-assets/blob/69846c4254ab4753782c06a24a7c00f1c93b0b6b/tapdb/sqlc/queries/authmailbox.sql#L52)):*
> ```TSQL
>     m.encrypted_payload,
>     m.arrival_timestamp,
>     m.expiry_block_height,
> ```

Ah so we're removing this everywhere after all in favor the the should-be-good-enough-for-everyone value? np, we can always re add it later in a migration if needed
___
*[`rpcserver.go` line 1689 at r11](https://reviewable.io/reviews/lightninglabs/taproot-assets/1587#-OWx56SK1Cf-Ndmi7qc8:-OWx56SK1Cf-Ndmi7qc9:b-8fx2u7) ([raw file](https://github.com/lightninglabs/taproot-assets/blob/a45379ab9ff0951cf1f5ed926dcaa5b00ad97984/rpcserver.go#L1689)):*
> ```Go
> 
> 	// If the user didn't specify a specific address version but is using
> 	// a group key, we default to version 2.
> ```

:+1: 
___
*[`proof/courier.go` line 1878 at r11](https://reviewable.io/reviews/lightninglabs/taproot-assets/1587#-OWx7_0u8K916gcpg0q-:-OWx7_0u8K916gcpg0q0:b-gpckrd) ([raw file](https://github.com/lightninglabs/taproot-assets/blob/a45379ab9ff0951cf1f5ed926dcaa5b00ad97984/proof/courier.go#L1878)):*
> ```Go
> 		// RPC methods that we require.
> 		statusErr, ok := status.FromError(err)
> 		if ok && statusErr.Code() == codes.Unimplemented {
> ```

Nice, I assume this is for cases where a Universe server hasn't updated to the latest version?
___
*[`cmd/commands/addrs.go` line 122 at r11](https://reviewable.io/reviews/lightninglabs/taproot-assets/1587#-OWx671yDUNykqFE_e-o:-OWx671yDUNykqFE_e-p:b-8fx2u7) ([raw file](https://github.com/lightninglabs/taproot-assets/blob/a45379ab9ff0951cf1f5ed926dcaa5b00ad97984/cmd/commands/addrs.go#L122)):*
> ```Go
> 		// at least a V2 address. So we can just use that if the user
> 		// specifies no explicit version.
> 		if len(groupKey) > 0 {
> ```

:+1: 


<!-- Sent from Reviewable.io -->

guggero reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/lightninglabs/taproot-assets/1587)* status: all files reviewed, 60 unresolved discussions (waiting on @ffranr, @Roasbeef, and @ZZiigguurraatt)

___
*[`itest/addrs_v2_test.go` line 50 at r11](https://reviewable.io/reviews/lightninglabs/taproot-assets/1587#-OWx72XU1ocgkECarV39:-OWyXtO-53q8l36bzfZS:b-x7swdg) ([raw file](https://github.com/lightninglabs/taproot-assets/blob/a45379ab9ff0951cf1f5ed926dcaa5b00ad97984/itest/addrs_v2_test.go#L50)):*
<details><summary><i>Previously, Roasbeef (Olaoluwa Osuntokun) wrote…</i></summary><blockquote>

What type of error would get user get if they attempted to send a value greater than 1 to a collectible address? I wonder if we should encode that extra bit into the addr type. 

FWIW, calls to decode the address _should_ fetch the genesis as needed allowing this full information to be displayed.
</blockquote></details>

Good question!
So due to this: https://github.com/lightninglabs/taproot-assets/blob/07d0c505127ea016736602731d6631542f19dcc1/address/address.go#L219
We don't allow addresses for collectibles to have a zero (optional) amount in the first place. So the error would already be thrown at address creation time: `unable to make new addr: unable to make new addr: address: collectible asset amount not one`.

I've added an integration test case for this!
___
*[`proof/courier.go` line 1878 at r11](https://reviewable.io/reviews/lightninglabs/taproot-assets/1587#-OWx7_0u8K916gcpg0q-:-OWyXpp23H_qdK3oOP74:b-9wdi06) ([raw file](https://github.com/lightninglabs/taproot-assets/blob/a45379ab9ff0951cf1f5ed926dcaa5b00ad97984/proof/courier.go#L1878)):*
<details><summary><i>Previously, Roasbeef (Olaoluwa Osuntokun) wrote…</i></summary><blockquote>

Nice, I assume this is for cases where a Universe server hasn't updated to the latest version?
</blockquote></details>

Yes, correct.
___
*[`tapdb/sqlc/queries/authmailbox.sql` line 52 at r6](https://reviewable.io/reviews/lightninglabs/taproot-assets/1587#-OWx7po1CsIdJkOEFT1Q:-OWyXgSr-OmbNvtf0KRU:b-1kogcw) ([raw file](https://github.com/lightninglabs/taproot-assets/blob/69846c4254ab4753782c06a24a7c00f1c93b0b6b/tapdb/sqlc/queries/authmailbox.sql#L52)):*
<details><summary><i>Previously, Roasbeef (Olaoluwa Osuntokun) wrote…</i></summary><blockquote>

Ah so we're removing this everywhere after all in favor the the should-be-good-enough-for-everyone value? np, we can always re add it later in a migration if needed
</blockquote></details>

Yes, exactly. We can always add something later once we know exactly how/when we can do a cleanup without affecting emergency recovery.


<!-- Sent from Reviewable.io -->
